File: tests/cxx/doc_snippets/test_modules.cpp
/*
 * Copyright 2022 NWChemEx-Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "electric_field.hpp"
#include "force.hpp"
#include "load_modules.hpp"
#include "modules.hpp"
#include <catch2/catch.hpp>

using namespace pluginplay_examples;
TEST_CASE("Tutorial modules") {
    // Set up module manager
    pluginplay::ModuleManager mm;
    // Load your modules
    load_modules(mm);
    REQUIRE(mm.count("Classical Force") == 1);
    REQUIRE(mm.count("Coulomb's Law") == 1);
    REQUIRE(mm.count("Coulomb's Law with screening") == 1);
    REQUIRE(mm.count("Single-precision Coulomb's law") == 1);
    REQUIRE(mm.count("Nonexistent module") == 0);

    auto force0_mod = mm.at("Classical Force");
    auto force1_mod = mm.at("Coulomb's Law");
    auto force2_mod = mm.at("Single-precision Coulomb's law");
    auto force3_mod = mm.at("Coulomb's Law with screening");
    Point p0{0.0, 0.0, 0.0};
    Point p1{1.0, 0.0, 0.0};
    Point p2{2.0, 0.0, 0.0};
    Point r0{5.5, 0.0, 0.0};
    Point r1{1.5, 0.0, 0.0};
    PointCharge pc0{1.0, p0};
    PointCharge pc1{1.0, p1};
    PointCharge pc2{2.0, p2};
    std::vector pcs{pc1, pc2};
    double m1{2.0};

    auto result0 = force0_mod.run_as<Force>(pc0, m1, p2, pcs);
    REQUIRE(result0 == r0);

    auto result1 = force1_mod.run_as<ElectricField>(p0, pcs);
    REQUIRE(result1 == r1);

    auto result2 = force2_mod.run_as<ElectricField>(p0, pcs);
    REQUIRE(result2 == r1);

    auto result3 = force3_mod.run_as<ElectricField>(p0, pcs);
    REQUIRE(result3 == r1);

    auto force3_mod_copy = force3_mod.unlocked_copy();
    force3_mod_copy.change_input("threshold", 1.0);
    auto screened_result3 = force3_mod_copy.run_as<ElectricField>(p0, pcs);
    REQUIRE(screened_result3 == p0);
}

File: tests/python/doc_snippets/test_modules.py
#
# Copyright 2023 NWChemEx-Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import parallelzone as pz
import unittest
import pluginplay as pp
import pluginplay_examples as ppe

class TestTutorialModules(unittest.TestCase):
    def test_modules(self):
        rv = pz.runtime.RuntimeView()
        mm = pp.ModuleManager()
        ppe.load_modules(mm)
        self.assertEqual(mm.size(),4)
        
        module_names = ["Coulomb's Law", "Classical Force", 
                        "Coulomb's Law with screening", 
                        "Single-precision Coulomb's law"]
        for name in module_names:
            self.assertTrue(mm.count(name))

        force0_mod = mm.at("Classical Force")
        force1_mod = mm.at("Coulomb's Law")
        force2_mod = mm.at("Single-precision Coulomb's law")
        force3_mod = mm.at("Coulomb's Law with screening")

        p0 = [0.0, 0.0, 0.0]
        p1 = [1.0, 0.0, 0.0]
        p2 = [2.0, 0.0, 0.0]
        r0 = [5.5, 0.0, 0.0]
        r1 = [1.5, 0.0, 0.0]
        pc0 = ppe.PointCharge(1.0, p0)
        pc1 = ppe.PointCharge(1.0, p1)
        pc2 = ppe.PointCharge(2.0, p2)
        pcs = [pc1, pc2]
        m1 = 2.0

        force_pt = ppe.Force()
        efield_pt = ppe.ElectricField()

        result0 = force0_mod.run_as(force_pt, pc0, m1, p2, pcs)
        self.assertTrue(result0 == r0)

        result1 = force1_mod.run_as(efield_pt, p0, pcs)
        self.assertTrue(result1 == r1)

        result2 = force2_mod.run_as(efield_pt, p0, pcs)
        self.assertTrue(result2 == r1)

        result3 = force3_mod.run_as(efield_pt, p0, pcs)
        self.assertTrue(result3 == r1)

        force3_mod_copy = force3_mod.unlocked_copy()
        force3_mod_copy.change_input("threshold", 1.0)
        screened_result3 = force3_mod_copy.run_as(efield_pt, p0, pcs)
        self.assertTrue(screened_result3 == p0)
